### ISSUE TITLE
Safari only bug: 'script-src' contains an invalid source: ''strict-dynamic''. It will be ignored.

### ISSUE THREAD

**Issue by mesqueeb** [google/google-api-javascript-client#397]
Dear GAPI team.
I have a security bug only on Safari.
However the GAPI script still works so this bug should be considered **low priority**.

Right in between loading and initialising GAPI I get these:

> [Error] The source list for Content Security Policy directive 'script-src' contains an invalid source: ''strict-dynamic''. It will be ignored.
> [Error] Refused to execute a script because its hash, its nonce, or 'unsafe-inline' does not appear in the script-src directive of the Content Security Policy.

<img width="831" alt="2018-02-21 10 34 04" src="https://user-images.githubusercontent.com/3253920/36458575-d9e77b84-16f2-11e8-8df5-3ec806c79d2b.png">

**Comment by TMSCH** [comment]
What version of Safari do you use?

**Comment by mesqueeb** [comment]
@TMSCH 11.0.3 (13604.5.6)

**Comment by TMSCH** [comment]
I can indeed see the message regarding `strict-dynamic`. It seems that Safari doesn't support this CSP directive. It doesn't seem to prevent the library to work though, so it's indeed low priority.
Thanks for the report!

**Comment by ksorokin** [comment]
Scalr/gantt#1

**Comment by EnVy28** [comment]
Yo! I have the same problem, tested in Safari 9.
 
@TMSCH said that it does not prevent the library to work, but somehow the sign out thingy is not working... I don't know if the error causes it or not though. Everytime I run my signout function it just reloads the page and not signing out.

`var auth2 = gapi.auth2.getAuthInstance();
auth2.signOut().then(function () {
    console.log('User signed out.');
    /* AJAX (Delete site session) */
});
auth2.disconnect();`

PS: It works on Chrome, IE 11, EDGE (All latest ver as of now)

**Comment by TMSCH** [comment]
@EnVy28 are you only observing the error mentioned in the first comment? Why are you disconnecting the user too? You should remove that call.

**Comment by EnVy28** [comment]
@TMSCH It's my first time using Google's SSO so I just followed the basic example written in the docs which is, exactly like the above code I posted except, there is no `auth2.disconnect();`.

But without `auth2.disconnect();`, somehow my application won't signout, it just redirects to the main page, even though I delete the users session in my app.

It's all good though, I stopped using MacOS since Lion so I didn't know that Safari 9 and 10 support are over so we just stopped supporting that too.

Currently there are no problems with Safari 11 so far, so it's all good...

**Comment by StonehengeCreations** [comment]
I get the same error on the latest version of Mas OSX with the latest version of Safari. So it appears to have not been solved.

**Comment by gp-birender** [comment]
yes this is happening in latest version of Mac safari

**Comment by ramiljoaquin-ubidy** [comment]
What is the solution for this issue?  We get the same error as well.

**Comment by abel672** [comment]
Hi, I am having this issue using the Google People API to the the gmail contacts.
Basically is throwing it when my code try to open the google api popup to signing with gmail. It doesn't open it by the way.

This is the code:

`gapi.load('client:auth2', () => {
            gapi.client.init({
                apiKey: 'MY_API_KEY',
                discoveryDocs: ['https://people.googleapis.com/$discovery/rest?version=v1'],
                clientId: 'MY_CLIENT_ID.apps.googleusercontent.com',
                scope: 'profile email https://www.googleapis.com/auth/contacts.readonly'
            }).then((data) => {
                // Signin with auth2 process (authentication from google)
                gapi.auth2.getAuthInstance().signIn()
                    .then((data) => {
                        console.log('data', data);
                        let signedin = gapi.auth2.getAuthInstance().isSignedIn.get();
                        // Once signed in, take data
                        if (signedin) {
                            console.log('people', gapi.client);
                            gapi.client.people.people.connections.list({
                                resourceName: 'people/me',
                                personFields: 'emailAddresses,names,phoneNumbers'
                            }).then((data) => {
                                let res = data.result;
                                console.log('data', res);
                                let phonesNumbers = res.connections;
                                let pageToken = res.nextPageToken;

                                // Displaying contacts
                                this.addContacts(phonesNumbers);

                                // NextPageToken
                                if (pageToken) {
                                    this.getMoreContacts(pageToken);
                                } else {
                                    this.showContacts();
                                }
                            }).catch((err) => {
                                console.log('err', err);
                            });
                        }
                        console.log('signedin', signedin);
                    }).catch((err) => {
                        console.log('Err', err);
                    });
            });
        });`

This is the error that is throwing Safari.

![screen_shot_2018-04-24_at_16 26 20](https://user-images.githubusercontent.com/22236662/39231886-d97bf9c2-486b-11e8-931c-29b584cf067b.png)

**Comment by mesqueeb** [comment]
@abel672 Are you sure you don't have an Adblock on? These usually block these.
In that case you should catch this specific error and display a message to the user asking to disable the adblock extension.

**Comment by akashzcoder** [comment]
I had the same issue, after disabling the adblocker, it started to work! Just curious, what does it have to do with the adblocker?

**Comment by mesqueeb** [comment]
@akashzcoder The browser thinks the login popup is an advertisement lol... That's either some bad coding by the AdBlock devs or I don't know if the Google Api Client can do anything to prevent such things.

**Comment by gp-birender** [comment]
I had the same issue but its working now, I have posted the serialized form and remove other ajax parameters.

**Comment by dohsimpson** [comment]
I have the same problem :(

For me, disabling ad blocker (ublock origin) doesn't work, but turning on incognito mode works.

**Comment by kolnogorov** [comment]
Did anybody solve an Original Poster's bug?
Is there any solution for this?

**Comment by abel672** [comment]
I solved it in this way thanks to a friend.
Open Safari and go to 'Preferences'
Go to 'Security' window and disable 'Block pop-up windows'.

![screen shot 2018-05-13 at 13 53 55](https://user-images.githubusercontent.com/22236662/39967030-38582022-56b5-11e8-8247-217aea7d93fb.png)

Hope that It can help.

**Comment by dohsimpson** [comment]
I get the following error in my Safari console after opening *google.com* for a few seconds. This causes the input prompt to lose focus, which is very annoying due to how often I use google.

<img width="1278" alt="screen shot 2018-05-19 at 7 23 31 pm" src="https://user-images.githubusercontent.com/5865309/40274005-2631eb14-5b9a-11e8-839d-1cbbf2dc8899.png">

After some digging, I realized that when the widget script for *notification.google.com* tries to execute, it will cause the error.

Solution: I added the following rule in my ad blocker (uBlock origin for Safari) to block the widget from loading.

```
www.google.com https://notifications.google.com/u/0/widget sub_frame block
```

<img width="1278" alt="screen shot 2018-05-19 at 7 29 11 pm" src="https://user-images.githubusercontent.com/5865309/40274034-edb1878a-5b9a-11e8-9ba0-3907457fbee8.png">

Hope this helps someone.

**Comment by marksdk** [comment]
For what it's worth, this is still an issue in Safari 12.0 (14606.1.22.2 - macOS Mojave 10.14 Beta (18A326g)), _without_ an adblocker installed.

**Comment by kkaliher** [comment]
We are seeing this error suddenly (as of maybe a couple days ago) causing firebase phone auth for iOS to fail

**Comment by mesqueeb** [comment]
@marksdk To me that's worth a lot. Thanks!! : )

**Comment by liquidvisual** [comment]
Thanks Safari!

**Comment by marksdk** [comment]
I kept looking into this issue for a while, and I ran into a lot of trouble with Chrome adhering to CSP v3 and Safari still adhering to CSP v2: https://www.chromestatus.com/features/5633814718054400. In short, nonce- or hash-based whitelists will be a requirement at some point, but right now the major browsers just interpret the same whitelists differently ¯\_(ツ)_/¯

That might help some of you narrow down the search when looking for an answer to CSP troubles.

**Comment by emolr** [comment]
I have scratched my head for a while but finally made it work for me. 

Instead of having `<script src="https://apis.google.com/js/platform.js"></script>` in my `index.html`, 
I'm appending it to `<head>` using Javascript.

```
const scriptNode = document.createElement('script');
scriptNode.src = this.gapiUrl;
scriptNode.type = 'text/javascript';
scriptNode.charset = 'utf-8';
document.getElementsByTagName('head')[0].appendChild(scriptNode);
```

I still get the error in safari console, but the Google API now works and I can log in users in safari even though it feels like the hackiest thing since targeting specific browser versions via CSS selectors.

**Comment by jeanpaul1304** [comment]
I have the same error, when my app uses google recaptcha in safari

**Comment by oprearocks** [comment]
> I have the same error, when my app uses google recaptcha in safari

Same thing here. Using ReCaptcha on Safari makes it freeze.

**Comment by therealcaii** [comment]
Getting the same error here using Invisible reCaptcha v2

**Comment by matviishyn** [comment]
Same error, using reCaptcha v2

**Comment by vbresults** [comment]
@abel672 thank you, now I at least know what's causing it.  Reproduced.

**Comment by vbresults** [comment]
Found the solution on SO.  You have to trigger the popup in response to a click.  There is no alternative and there likely never will be unless google cracks Safari's popup blocker.

@abel672 thank you for informing us of the cause, it helped find the answer.

**Comment by yuruxuan** [comment]
The same error, but it appear today first time : (
Safari 11.1.2

**Comment by anishkny** [comment]
I am seeing this in [http://console.firebase.google.com](http://console.firebase.google.com) on Safari. The console fails to load 😢

**Comment by yuruxuan** [comment]
> I am seeing this in http://console.firebase.google.com on Safari. The console fails to load 😢

I found just http://console.firebase.google.com can not load. You can visit your project page directly, like https://console.firebase.google.com/project/your project id/overview

**Comment by anishkny** [comment]
Update: http://console.firebase.google.com/ started working for me earlier this week. No changes to Safari still Version 11.0.3 (13604.5.6). macOS High Sierra 10.13.3.

**Comment by tjkohli** [comment]
I still have this error using Google reCaptcha on Safari 12.0.2.

**Comment by harshadduasane** [comment]
Hello,
i am also facing same issue. tried almost all solution provided above but no success.

Any one found solution? please help

**Comment by surger** [comment]
Also have a troubleshooting with google recaptcha on Safari

**Comment by saramoran** [comment]
Hi, 
I also have this problem using Recaptcha (2 or 3, I tried both) on Safari 12.0.3.

**Comment by umpc** [comment]
"It doesn't seem to prevent the library to work"

But it creates false alarms on production sites that use APM tools.

The actual error messages are hidden by default, due to crossorigin rules, so I only found out that this was the cause after real-device testing.

Any fix for this issue on ReCAPTCHA would be greatly appreciated.

**Comment by Lozik** [comment]
I can confirm that issue. I'm on Safari 12.0.3 on MacOs 10.13.6. No addblockers installed.
Is Recaptcha even using this client? It does not figure in the [listing of the README](https://developers.google.com/apis-explorer/#p/).

Even on the [Google Recaptcha demo page](https://www.google.com/recaptcha/api2/demo), the error is present. See screenshots.

This is extremely painful for users using Safari.

<img width="1439" alt="Screen Shot 2019-03-11 at 20 07 12" src="https://user-images.githubusercontent.com/16655002/54150674-51efb780-4439-11e9-8260-7cf9e1160c87.png">

**Comment by existential-fallacy** [comment]
Not every browser, even their respective latest versions, supports **every** CSP directive. Period.

**Comment by cleathersden** [comment]
sooo... there's no solution ?

**Comment by cleathersden** [comment]
Try adding "localhost" to the recaptcha list of domains.

**Comment by grant** [comment]
Hi all,
I believe this issue is due to Safari being a browser without CSP 3.0 compliance and thus will ignore `strict-dynamic`.

Here's an example site you can view to see compliance:
https://csp-experiments.appspot.com/strict-dynamic

Related tracking issue:
https://github.com/Fyrd/caniuse/issues/2315

Progress on this issue is blocked by that Safari support.

**Comment by powolnymarcel** [comment]
Hi, 

I had the same problem, I added a nonce attr  to my script tag and my project is working on Safari... 
`
var scriptElement = document.createElement('script');
`
`
scriptElement.nonce = 'this_is_my_nonce';
` 
`
scriptElement.setAttribute('recaptcha-v3-script', '');
`
`
var scriptBase = 'https://www.google.com/recaptcha/api.js';
`

**Comment by PhamLeHai** [comment]
For my case, even after loading the api.js by javascript, no error log in console, but once it tries to reloadAuthResponse(), it will receive an error 'userLoggedOut', I can not find any document/hint about the error.

Suddenly, after I changed the UX mode of the .signIn() function from 'redirect' to default (popup), there are still errors on the console, but magically, the script can load and users can log in normally.

I still don't understand the reason, since the code of the auth2 module is converted and hard to read on the browser. Anyway, it seems working. Hope this can help others. Please let me know if there is anyway to dig deeper about this issue.

**Comment by tvb** [comment]
I got the same problem at MacOS 10.15.2 (19C57) and Safari Version 13.0.4 (15608.4.9.1.3).

**Comment by morscad** [comment]
Hi All

We have the same issue making calls to the google reaptcha v3 API from a web application (React + Gatsby) which is loaded in a react-native webview on iOS.

Safari on desktop (MacOS 10.15.2) allows the call to pass but with lots of warnings, but react-native-webview on iOS blocks the call entirely.
the same code works fine on Android, and on Chrome and Firefox on desktop.

anyone had any breakthroughs in making this happen?

**Comment by mamallama** [comment]
Is there a solution to this? Having the same issue, no popup blocker, MacOS 10.15.2,  Safari version 13.0.4, only error: 
"The source list for Content Security Policy directive 'script-src' contains an invalid source: ''strict-dynamic''. It will be ignored."

It's keeping me from submitting a form.

**Comment by paulbeard** [comment]
`Refused to execute a script because its hash, its nonce, or 'unsafe-inline' does not appear in the script-src directive of the Content Security Policy.
`

is also showing on GMail's web UI. The old 2004 vintage/low or slow bandwidth version works fine, but the new Standard view doesn't load. It's been balking for a few days, but now just fails.

**Comment by kamzata** [comment]
I got the same error using it on Magento 2. Any solution?

**Comment by joshbruce** [comment]
Same - using raw via PHP + Laravel (no packages).

**Comment by fabiosoto** [comment]
Same - Safari 12.1.2, Angular + ReCaptcha

**Comment by dinohorvat** [comment]
Same issue. Safari 13.1.1, Angular + Trading View Widget

**Comment by ArjenB** [comment]
Same here using Divi theme's implementation of  Recaptcha 3 in Wordpress on Safari 13.1.1 on MacOS Catalina 10.15.6. However this doesn't prevent my Contact Form to be submitted on any device (Mac, iPhone, iPad). Just looks sloppy to have so many errors in the console.

**Comment by gregblass** [comment]
I have rollbar (an error reporting tool) hooked up to my javascript in production sites. I take errors seriously as a developer. This is such a shame.

**Comment by mesqueeb** [comment]
Here I stumble back on the same issue almost 3 years later.

This time I cannot seem to use Google reCaptcha in a **Capacitor app for iOS**.

**Comment by kirilltitov** [comment]
Forget it, they just don't care.

**Comment by PierreSymonPeralta** [comment]
Same here.

`The source list for Content Security Policy directive 'script-src' contains an invalid source: ''strict-dynamic''. It will be ignored.`

Google ReCaptcha v2 invisible
Safari Version 13.1.1 (15609.2.9.1.2)

**Comment by kinucris** [comment]
I have the same problem with _Safari 14.0 15610.1.28.1.9_ 
with reCaptcha too

**Comment by THammond9** [comment]
This error prevents our form from loading in Safari and iOS

**Comment by sergentj** [comment]
Hi all,

I see a lot of activity here and I'd like to be able to help, but it's not clear to me how. Here are a few questions I would like to know the answer to, from anyone who is still having trouble.

1. Is there is an issue with gapi as opposed to reCAPTCHA? This bug tracker is for gapi, not reCAPTCHA; as far as I can tell the recent complaints over the last year or two are about reCAPTCHA and not gapi. If there is some integration between them that I'm not aware of, I apologize, but maybe you can point me to what it is. This bug was originally filed because gapi was triggering these warnings, but that was fixed years ago as far as I know.

From what I can tell, the reCAPTCHA team uses [StackOverflow](https://stackoverflow.com/questions/tagged/recaptcha)
for helping out on questions like this, so if the problem is purely reCAPTCHA related I suggest posting there instead.

2. Are the error messages are just spurious warnings or do they actually indicate breakage? reCAPTCHA's demo does seem to work on Safari, so I'm not sure if the error messages are serious or are just ignorable log spam. I also see a [few](https://stackoverflow.com/questions/55704552/content-security-policy-issue-in-google-recaptcha-v2) stack overflow [answers](https://stackoverflow.com/a/61746213) suggesting that they are just notices and not fatal errors.

Can anyone who is having trouble clarify answers to these two questions for their use case? Thanks.

**Comment by tom-b-wright** [comment]
Hi @sergentj 

Since your team already solved a similar issue years ago, I think communicating the solution to the recaptcha team will be the best approach.

We are getting this error message in the console (Safari): **The source list for Content Security Policy directive 'script-src' contains an invalid source: ''strict-dynamic''. It will be ignored.**

I believe that the "strict-dynamic" header value is transmitted by the server when issuing a GET HTTP request to https://www.google.com/recaptcha/api2/anchor

So the task is to check why that HTTP response includes that header value and whether it is possible to make it supported in Safari.
This is a sample value of that header:
Content-Security-Policy: script-src 'nonce-Y2U53p7Dpty6XSdv8PgARw' 'unsafe-inline' 'strict-dynamic' https: http: 'unsafe-eval';object-src 'none';base-uri 'self';report-uri https://csp.withgoogle.com/csp/recaptcha/1

Thanks.

**Comment by sergentj** [comment]
Hi, thanks for your reply. It'd be better if you can engage directly with the reCAPTCHA team for help on this rather than use me as a go-between, because I'm sure they will have followup questions for you. According to their documentation, you should email support@recaptcha.net for bugs and feature requests, or use [Stack Overflow](https://stackoverflow.com/questions/tagged/recaptcha?sort=frequent) for general support.

**Comment by egranty** [comment]
![](https://user-images.githubusercontent.com/3253920/36458575-d9e77b84-16f2-11e8-8df5-3ec806c79d2b.png)
these are 2 independent error/warning messages those appear in Safari [independently](https://csplite.com/csp/test214/#safari) each other.
Special thanks to @mesqueeb for a detailed data for research provided.
So, let's examine each one separately.

1. *[Error] The source list for Content Security Policy directive 'script-src' contains an invalid source: ''strict-dynamic''. It will be ignored.*
Safari [doesn't fully support 'strict-dynamic'](https://csplite.com/csp/test217/#safari) and honestly warns about it in its console. Actually it's not an error, it's just a warning. Note that this error does not contain the name of the affected script, this is a general warning.
Such warnings are normal practice for all browsers.
We will have to live with this warning until Safari is fully CSP3 compliant.

Anyway, the Google API Client Library developers have no control over Apple's state of affairs.

2. [Error] Refused to execute a script because its hash, its nonce, or 'unsafe-inline' does not appear in the script-src directive of the Content Security Policy.
The err means some inline script is really locked. Why this does not prevent the Google API from normal operate, see para 4 below.

3. In 99% of cases, this error occurs inside an invisible Google frame, in which data is exchanged with Google services. Pay attention to the place of error occurred: postmessageRelay (it works in its frame) and iframe: 0
To protect this frame, related Google service publishes within it the own CSP in [browsers backward compatibility mode](https://csplite.com/csp198/#CSP_backward_compatibility), one of such CSP is shown by @tom-b-wright above.
Do reduce security of auth iframes Google just because of Safari console warnings - not a good idea.

1% of cases:
- @EnVy28 : looks like he has problem with X-Frame-Options header, it is a separate [Safari bug](https://csplite.com/csp17/#bug_Safari_frame-ancestors_X-Frame-Options)
- @abel672  does not have a second error message, looks like  Safari just locks a popup window due to some other reason.
If such error related to you own CSP, not Google's one, it could be [fixed](https://csplite.com/csp155/#v3) in some cases.

4. Why reCAPTCHA/Auth2 continues to operate even second error occurs and some calls seems to be blocked?
- I think in some cases Google uses backup way to delivery his requests, because scripts are created on browsers compatibility manner.
- most common situation is an usage <a href='javascript:void(0)'> construct to prevent following the hyperlink. CSP locks it as inline script, but simultaneously it prevents follow link the same as void(0).

Safari is special. Therefore developers have a related headache like it was with IE6 few years ago.

**Comment by smartameer** [comment]
<img width="904" alt="image" src="https://user-images.githubusercontent.com/5998873/100279952-72550200-2f8d-11eb-8c68-9c4624464438.png">
Uncheck the *Prevent cross-site tracking* solved my issue of cross site images loading. ( MacOS Big Sur )

**Comment by XvJX** [comment]
Mac
![WX20201127-163628](https://user-images.githubusercontent.com/29369288/100428603-2316ea80-30cf-11eb-9e35-e7aad742f56f.png)
IPhone
![1606466321109](https://user-images.githubusercontent.com/29369288/100428647-2e6a1600-30cf-11eb-8a5c-44f3bee02785.jpg)
I hope it can help you！

**Comment by TukuToi** [comment]
The solution to remove block pop ups setting is not a solution, it more a workaround/hack, and additionally doesn't solve a thing for me.

I can't believe this is being ignored? Concurrence Google/Apple being the issue, or what is the reason a 3 years old BUG is not resolved?

Any other solution than disabling blockers?

**Comment by veeralpatel** [comment]
Are there any updates here?

**Comment by danielraban** [comment]
Are we going to get a solution?

**Comment by TukuToi** [comment]
Don't we worry - google is busy destroying our all businesses with their "core web vitals" updates and invading our all privacies with things like FLoC, while things like [these](https://cognitiveseo.com/blog/24578/google-indexing-problem/) or just the very GIT Thread here... proof how much of a authoritative instance Google is for things "good web":
None!

In fact it is their broken tools, indexes not working or badly working, and other issues clearly show how much they are interested on _your success_ versus _their success by gathering data for free thru their analytics services_

How did Google even become this hilarious monopoly giant? In other industries, they would be investigated for monopoly tactics and probably broken into pieces, if not sued for favouritism.

Going off topic, however, it is kind of on topic, since it all concerns the one giant PITA we deal with here:
google

**Comment by sreeram-solinst** [comment]
Any updates? I'm unable to load reCaptcha with Safari v14.0!

**Comment by pcarballeda** [comment]
I am experiencing the same problem with Recapcha invisible and Safari.

**Comment by xKZL** [comment]
Same here ! :(

**Comment by kopkako** [comment]
Running into this as well, sadly did not find any solution other than having to disable ReCaptcha (v3, invisible) for the time being. Is there any update on this? Could anyone from Google please help us out, here?

As it stands right now, apparently ReCaptcha simply does not work on Safari? Does no one care about this?

**Comment by kirilltitov** [comment]
Given that they make their products virtually unusable for non-Chrome browsers (Meet, for example), it's not like they don't care, but rather, I'm sure, it is intended behaviour.

**Comment by dvanderb** [comment]
This looks like a question about Google Login and Authentication. Please address this question to the google-signin tag on stackoverflow(https://stackoverflow.com/questions/tagged/google-signin)

If you believe this is in error, please reopen the issue.
